File: goonam_manage_app/lib/models/order.dart
import 'stage.dart';
import 'comment.dart';

class Order {
  final int id;
  final String orderNo;
  final String projectName;
  final String? companyName;
  final String orderType;
  final String? description;
  final int quantity;
  final DateTime orderDate;
  final DateTime deliveryDate;
  final DateTime? originalDeliveryDate;
  final DateTime? actualEnd;
  final String status; // trạng thái hiện tại
  final String overallStatus; // tổng quan: in_progress / late / done
  final int highlightFlag;
  final String? timeBudgetDays; // đổi sang String để dễ hiển thị
  final String? remainingDays; // đổi sang String để dễ hiển thị
  final List<Stage> stages;
  final List<Comment> comments;
  final List<EditHistory> history; // lịch sử chỉnh sửa đơn hàng
  final List<dynamic>? adminEditLogs; // ghi lại log chỉnh sửa của admin
  final DateTime? deletedAt;
  final bool isPaused;
  final String? pauseReason;
  final DateTime? pausedAt;
  final DateTime? completedAt;
  final Map<String, dynamic>? latestComment;
  final int unreadCommentsCount;
  Order({
    required this.id,
    required this.orderNo,
    required this.projectName,
    this.companyName,
    required this.orderType,
    this.description,
    required this.quantity,
    required this.orderDate,
    required this.deliveryDate,
    this.originalDeliveryDate,
    this.actualEnd,
    required this.status,
    required this.overallStatus,
    required this.highlightFlag,
    this.timeBudgetDays,
    this.remainingDays,
    required this.stages,
    this.comments = const [],
    this.history = const [],
    this.adminEditLogs,
    this.deletedAt,
    this.isPaused = false,
    this.pauseReason,
    this.pausedAt,
    this.completedAt,
    this.latestComment,
    this.unreadCommentsCount = 0,
  });

  /// ✅ Parse từ JSON và KHÔNG bị lệch ngày (giữ nguyên theo local VN)
  factory Order.fromJson(Map<String, dynamic> json) {
    // ✅ HÀM CŨ: Parse ngày (chỉ lấy yyyy-MM-dd, bỏ giờ)
    DateTime _parseLocalDate(String? s) {
      if (s == null || s.trim().isEmpty || s == "0" || s.toLowerCase() == "null") {
        return DateTime.now();
      }
      try {
        if (RegExp(r'^\d{4}-\d{2}-\d{2}$').hasMatch(s)) {
          final parts = s.split('-');
          return DateTime(
            int.parse(parts[0]),
            int.parse(parts[1]),
            int.parse(parts[2]),
          );
        }

        final parsed = DateTime.parse(s);
        final local = parsed.toLocal();
        return DateTime(local.year, local.month, local.day);
      } catch (e) {
        return DateTime.now();
      }
    }

    // ✅ HÀM MỚI: Parse datetime (giữ nguyên giờ + timezone)
    DateTime? _parseDateTime(String? s) {
      if (s == null || s.trim().isEmpty || s == "0" || s.toLowerCase() == "null") {
        return null;
      }
      try {
        final parsed = DateTime.parse(s); // Parse (tự động về UTC)
        return parsed.toLocal(); // ✅ CHUYỂN VỀ LOCAL TIMEZONE
      } catch (e) {
        return null;
      }
    }

    return Order(
      id: json['id'] ?? 0,
      orderNo: json['order_no'] ?? '',
      projectName: json['project_name'] ?? '',
      companyName: json['company_name'],
      orderType: json['order_type'] ?? '',
      description: json['description'],
      quantity: json['quantity'] is int
          ? json['quantity']
          : int.tryParse(json['quantity']?.toString() ?? '0') ?? 0,
      orderDate: _parseLocalDate(json['order_date']),
      deliveryDate: _parseLocalDate(json['delivery_date']),
      originalDeliveryDate: json['original_delivery_date'] != null
          ? _parseLocalDate(json['original_delivery_date'])
          : null,
      actualEnd: json['actual_end'] != null
          ? _parseLocalDate(json['actual_end'])
          : null,
      status: json['status'] ?? '',
      overallStatus: json['overall_status'] ?? '',
      highlightFlag: json['highlight_flag'] is int
          ? json['highlight_flag']
          : int.tryParse(json['highlight_flag']?.toString() ?? '0') ?? 0,
      timeBudgetDays: json['time_budget_days']?.toString(),
      remainingDays: json['remaining_days']?.toString(),
      stages: (json['stages'] is List)
          ? (json['stages'] as List)
          .map((e) => Stage.fromJson(e as Map<String, dynamic>))
          .toList()
          : [],
      comments: (json['comments'] as List<dynamic>?)
          ?.map((e) => Comment.fromJson(e as Map<String, dynamic>))
          .toList() ??
          [],
      history: (json['history'] as List<dynamic>?)
          ?.map((e) => EditHistory.fromJson(e as Map<String, dynamic>))
          .toList() ??
          [],
      adminEditLogs: json['admin_edit_logs'] ?? [],
      deletedAt:  _parseDateTime(json['deleted_at']),
      isPaused: json['is_paused'] ?? false,
      pauseReason: json['pause_reason'],
      pausedAt: json['paused_at'] != null
          ? _parseDateTime(json['paused_at']) // ✅ DÙNG HÀM MỚI
          : null,
      completedAt: _parseDateTime(json['completed_at']),
      // ✅ THÊM DÒNG NÀY
      latestComment: json['latest_comment'] as Map<String, dynamic>?,
      unreadCommentsCount: json['unread_comments_count'] ?? 0,
    );
  }

  /// ✅ Gửi ngược về backend: luôn theo dạng yyyy-MM-dd (đúng chuẩn Laravel)
  Map<String, dynamic> toJson() {
    String _formatDate(DateTime d) =>
        "${d.year.toString().padLeft(4, '0')}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}";

    return {
      'id': id,
      'order_no': orderNo,
      'project_name': projectName,
      'company_name': companyName,
      'order_type': orderType,
      'description': description,
      'quantity': quantity,
      'order_date': _formatDate(orderDate),
      'delivery_date': _formatDate(deliveryDate),
      'original_delivery_date': originalDeliveryDate != null
          ? _formatDate(originalDeliveryDate!)
          : null,
      'actual_end': actualEnd != null ? _formatDate(actualEnd!) : null,
      'status': status,
      'overall_status': overallStatus,
      'highlight_flag': highlightFlag,
      'time_budget_days': timeBudgetDays,
      'remaining_days': remainingDays,
      'stages': stages.map((s) => s.toJson()).toList(),
      'comments': comments.map((c) => c.toJson()).toList(),
      'history': history.map((h) => h.toJson()).toList(),
      'admin_edit_logs': adminEditLogs,
      'is_paused': isPaused,
      'pause_reason': pauseReason,
      'paused_at': pausedAt?.toIso8601String(),
    };
  }
}

/// ✅ Lớp ghi lại lịch sử chỉnh sửa đơn hàng
class EditHistory {
  final String editor;
  final String change;
  final DateTime date;
  final String? changeMode; // admin_edit, manual, system
  final String? note; // ghi chú tùy chọn

  EditHistory({
    required this.editor,
    required this.change,
    required this.date,
    this.changeMode,
    this.note,
  });

  factory EditHistory.fromJson(Map<String, dynamic> json) {
    DateTime _parseLocal(String? s) {
      if (s == null || s.isEmpty) return DateTime.now();
      try {
        final parsed = DateTime.tryParse(s);
        if (parsed == null) return DateTime.now();
        final local = parsed.toLocal();
        return DateTime(local.year, local.month, local.day);
      } catch (e) {
        return DateTime.now();
      }
    }

    return EditHistory(
      editor: json['editor'] ?? 'Không rõ',
      change: json['change'] ?? '',
      date: _parseLocal(json['date']),
      changeMode: json['change_mode'],
      note: json['note'],
    );
  }

  Map<String, dynamic> toJson() {
    String _formatDate(DateTime d) =>
        "${d.year.toString().padLeft(4, '0')}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}";

    return {
      'editor': editor,
      'change': change,
      'date': _formatDate(date),
      'change_mode': changeMode,
      'note': note,
    };
  }
}

File: goonam_manage/app/Models/Order.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;

class Order extends Model
{
    use HasFactory;
    use SoftDeletes;
    protected $fillable = [
        'order_type',
        'project_name',
        'company_name',
        'order_no',
        'description',
        'quantity',
        'order_date',
        'delivery_date',
        'original_delivery_date',
        'actual_end',
        'status',
        'history',
        'is_paused',
        'pause_reason',
        'paused_at',
        'completed_at',
    ];

    protected $casts = [
        'id' => 'integer',
        'quantity' => 'integer',
        'highlight_flag' => 'integer',
        'order_date'    => 'date',
        'delivery_date' => 'date',
        'original_delivery_date' => 'date',
        'actual_end'    => 'date',
        'history'       => 'array',
        'admin_edit_logs' => 'array',
        'is_paused' => 'boolean',
        'paused_at' => 'datetime',
        'completed_at' => 'datetime',
    ];

    protected $appends = [
        'overall_status',
        'remaining_days',
        'time_budget_days', // ✅ bỏ phần 'history' => 'array'
    ];
    protected $dates = ['deleted_at'];
    // ======= ACCESSORS =======
 public function getOverallStatusAttribute()
{
    $original = $this->original_delivery_date ?? $this->delivery_date;
    $actual   = $this->actual_end;
    $now      = now()->startOfDay();

    // ✅ Lấy các công đoạn có thời gian
    $realStages = $this->stages->filter(function ($s) {
        return !$s->is_skipped
            && (!is_null($s->planned_start) || !is_null($s->planned_end));
    });

    // ✅ TH1: Nếu KHÔNG có công đoạn nào có thời gian
    if ($realStages->isEmpty()) {
        if ($this->status === 'completed') {
            return 'done';
        }

        // ✅ Chỉ tính trễ khi ĐÃ QUA ngày giao (> endOfDay)
        if ($original) {
            $deliveryDay = \Carbon\Carbon::parse($original)->endOfDay(); // 23:59:59
            if ($now->gt($deliveryDay)) { // Chỉ > (không bao gồm =)
                return 'late';
            }
        }

        return 'in_progress';
    }

    // ✅ TH2: Nếu tất cả công đoạn đã hoàn thành
    if ($realStages->every(fn($s) => $s->status === 'done')) {
        if (!$actual) {
            return 'completed_early';
        }

        if ($original) {
            $originalDay = $original->copy()->startOfDay();
            $actualDay = $actual->copy()->startOfDay();

            if ($actualDay->gt($originalDay)) {
                return 'completed_late';
            } elseif ($actualDay->lt($originalDay)) {
                return 'completed_early';
            }
        }

        return 'done';
    }

    // ✅ TH3: Kiểm tra công đoạn trễ (> endOfDay)
    foreach ($realStages as $stage) {
        if ($stage->status === 'done') continue;

        if ($stage->planned_end) {
            $deadline = \Carbon\Carbon::parse($stage->planned_end)->endOfDay();

            if ($now->gt($deadline)) { // Chỉ > (không bao gồm =)
                return 'late';
            }
        }
    }

    // ✅ TH4: Nếu đã qua ngày giao hàng (> endOfDay)
    if ($original) {
        $originalDay = $original->copy()->endOfDay();
        if ($now->gt($originalDay)) { // Chỉ > (không bao gồm =)
            return 'late';
        }
    }

    return 'in_progress';
}

    public function getTimeBudgetDaysAttribute()
    {
        if (!$this->order_date || !$this->delivery_date) return null;

        $minutes = $this->order_date->diffInMinutes($this->delivery_date, false);
        if ($minutes <= 0) return "0 giờ";

        $days = intdiv($minutes, 1440);
        $hours = intdiv($minutes % 1440, 60);

        $result = '';
        if ($days > 0) $result .= $days . ' ngày';
        if ($hours > 0) $result .= ($result ? ' ' : '') . $hours . ' giờ';

        return $result ?: 'Dưới 1 giờ';
    }

    public function getRemainingDaysAttribute()
{
    if (!$this->delivery_date) return null;

    $now = now();
    $deadline = $this->delivery_date->copy()->endOfDay(); // ✅ Lấy cuối ngày

    $minutes = $now->diffInMinutes($deadline, false);
    $isLate = $minutes < 0;

    // ✅ Nếu còn trong ngày deadline → không hiển thị số âm
    if ($isLate && $now->isSameDay($this->delivery_date)) {
        return "0 giờ"; // Hoặc "Hết hạn hôm nay"
    }

    $absMinutes = abs($minutes);
    $days = intdiv($absMinutes, 1440);
    $hours = intdiv($absMinutes % 1440, 60);

    $result = '';
    if ($days > 0) $result .= $days . ' ngày';
    if ($hours > 0) $result .= ($result ? ' ' : '') . $hours . ' giờ';
    if ($result === '') $result = 'Dưới 1 giờ';

    return $isLate ? "-$result" : $result;
}

    // ======= RELATIONS =======
    public function stages()
    {
        return $this->hasMany(OrderStage::class)->orderBy('sequence');
    }

    public function comments()
    {
        return $this->hasMany(Comment::class);
    }

}
